### PR TITLE
Make fields optional for App and AppCatalog CRDs

### PR DESCRIPTION
**Pull request by rossf7**
Towards https://github.com/giantswarm/giantswarm/issues/11556

See https://gigantic.slack.com/archives/CR99LSZ1N/p1592311735122800. 

Mea culpa I think I approved the PR 🤦‍♂️ but the current openapi validation is too strict and isn't helping.

## Checklist

- [x] Consider SIG UX feedback.
- [x] Update changelog in CHANGELOG.md.
- [ ] If adding a new type, ensure that it is added to the [CRD unit tests](https://github.com/giantswarm/apiextensions/blob/d78aba91d578d56ef49f58e336035d35edc4e1b0/pkg/crd/crd_test.go#L9).

### CODE REVIEW

File: config/crd/v1/application.giantswarm.io_appcatalogs.yaml [hunk 1]
```diff
@@ -73,9 +73,6 @@ spec:
                     - name
                     - namespace
                     type: object
-                required:
-                - configMap
-                - secret
```
Comment (rossf7) [inline config/crd/v1/application.giantswarm.io_appcatalogs.yaml:78]: It's valid for an appcatalog to have no configuration. I think that applies to the releases catalog.

File: config/crd/v1/application.giantswarm.io_apps.yaml [hunk 2]
```diff
@@ -115,9 +112,7 @@ spec:
                     - namespace
                     type: object
                 required:
-                - context
                 - inCluster
-                - secret
```
Comment (rossf7) [inline config/crd/v1/application.giantswarm.io_apps.yaml:120]: Only inCluster is required. For CP app CRs we set this to true.

File: config/crd/v1beta1/application.giantswarm.io_apps.yaml [hunk 3]
```diff
@@ -76,9 +76,6 @@ spec:
                   - name
                   - namespace
                   type: object
-              required:
-              - configMap
-              - secret
```
Comment (rossf7) [inline config/crd/v1beta1/application.giantswarm.io_apps.yaml:81]: Same for app CRs. Its valid to have no config or userConfig.

File: pkg/apis/application/v1alpha1/app_catalog_types.go [hunk 4]
```diff
@@ -45,7 +45,7 @@ func NewAppCatalogCR() *AppCatalog {
 
 type AppCatalog struct {
 	metav1.TypeMeta   `json:",inline"`
-	metav1.ObjectMeta `json:"metadata"`
```
Comment (rossf7) [inline pkg/apis/application/v1alpha1/app_catalog_types.go:48]: Same change as https://github.com/giantswarm/apiextensions/pull/492.

Using omitempty for the optional fields is a lot nicer when using kubectl IMO.

**Review by marians — APPROVED** [review]

**Review by tomahawk28 — COMMENTED** [review]
From the last PR https://github.com/giantswarm/apiextensions/pull/488/files, I learn that `nullable` property is also necessary.

**Comment by rossf7** [comment]
> From the last PR https://github.com/giantswarm/apiextensions/pull/488/files, I learn that nullable property is also necessary.

@tomahawk28 Thanks good catch. I'll get that fixed.

**Comment by rossf7** [comment]
@tomahawk28 Should be good now. PTAL

**Review by tomahawk28 — APPROVED** [review]

**Comment by tomahawk28** [comment]
Sorry for late, LGTM!

**Comment by tomahawk28** [comment]
I found a PR a year ago.
> The problem is the operator changes got big and it needs logic to handle nils which is easy to miss
https://github.com/giantswarm/apiextensions/pull/285

I think we don't have those cases to miss anymore, is that right?

**Comment by rossf7** [comment]
> Sorry for late, LGTM!

Np :) ❤️ 

Tested on `geckon`.

**Comment by rossf7** [comment]
> I found a PR a year ago.

> The problem is the operator changes got big and it needs logic to handle nils which is easy to miss
#285

@tomahawk28 Thanks for posting this. I think we do need these changes because for example `.Spec.Config` can be nil when getting `.Spec.Config.ConfigMap.Name`.

But now that app CRs are customer facing I think it's important we get rid the empty string fields. So I'll re-implement the changes in https://github.com/giantswarm/app-operator/pull/136

**Pull request merged**